Plutarch/Integer.hs

```diff
@@ -7,6 +7,18 @@ import qualified PlutusCore as PLC
 
 data PInteger s
 
+class PIntegral a where
+  pdiv :: Term s a -> Term s a -> Term s a
+  pmod :: Term s a -> Term s a -> Term s a
+  pquot :: Term s a -> Term s a -> Term s a
+  prem :: Term s a -> Term s a -> Term s a
```

Comment:
You probably want to make these Plutarch-level functions, since e.g. `Rational` would probably hoist them. It's a trade-off though.

Reply:
Another place where I debated what I should've done. But I think it might not matter for just applying builtin functions? Correct me if I am wrong but if these were Plutarch level functions - applying them with `#` would still create an `Apply` term for each application, right?

Reply:
It matters if you want to pass the functions to a higher order function. Wonder whether there's any way we could eliminate this headache though.

Reply:
In the meantime, let's make these Plutarch level functions? I don't quite see a problem with that here.

Are you thinking of eta reductions? A highly simplified implementation perhaps? Just to get rid of cases where we have a term `X`, which is a `Lam` holding just an `Apply`. And then if someone puts the `X` term inside an `Apply` again - perhaps we can just cut the middle man and only yield the innermost `Apply`?

Reply:
We essentially have two types of functions: Haskell-level functions and Plutarch-level functions.
Given a Haskell-level function `f`, `plam f` converts it to a Plutarch-level function.
To go the opposite direction, there is no shortcut, but we essentially do `\v1 ... vn -> (plam f) # v1 # ... # vn`.
A round-trip always makes the term bigger.
We can not naively reduce `plam \x -> (plam f) # x` through eta-reduction, as the language has effects (`perror`).
This code compiles down into something like `\x -> f' x`, which if reduced to `f'`, might cause the term to err if `f'` is not in WHNF.

Let's say somebody does
```haskell
f x = f' # x
```
Then somebody wants to pass it to e.g. `pmap :: Term s ((a :--> b) :--> PList a :--> PList b)`, and does the following:
```haskell
pmap # (plam f) # l
```
`plam f` would be compiled into something like the following:
```haskell
\x -> f' x
```
This is often suboptimal. In the case where `f'` is a variable (e.g. a hoisted term), then it's already in WHNF and thus we can eta-reduce it without changing the semantics. However, this becomes *much* more complicated in functions of arity above 1:
```haskell
f' :: Term s (PInteger :--> PInteger :--> PInteger)
f x y = f' # x # y
```
In this case `plam f` would be compiled into:
```haskell
\x -> \y -> f' x y
```
Even if `f'` is in WHNF, the simple algorithm doesn't work anymore as `f' x` isn't in WHNF obviously.

Regardless of this, we still want to be able to do this transformation manually, by *trusting* the user.
The transformation we want to do is essentially `\f -> plam $ \x -> plam f # x` to just `\f -> plam f`.
This transformation is always safe, however, the inner `plam f` could in fact be a reference to a hoisted
term or similar, which is what complicates this.

If we had `plamReduce` which would eta-reduce its body, we could selectively do this.

An example this would fix:
```haskell
x + y = punsafeBuiltin PLC.AddInteger # x # y

padd = plam (+) -- Compiled into `\x -> \y -> AddInteger x y`
padd' = plamReduce (+) -- Compiled into `AddInteger`
```

An example where this would go wrong:
```haskell
_padd :: Term s (PInteger :--> PInteger :--> PInteger)
_padd = perror

x + y = _padd # x # y

padd = plam (+) -- Compiled into `\x -> \y -> error`
padd' = plamReduce (+) -- Compiled into `error`, bad
```

This is potentially surprising behaviour.

Perhaps the solution is just to perform deep analysis of terms.

We could special-case hoisted terms by adding another guarantee:
- Hoisted terms must not inspect their arguments before all arguments
  are passed.

This would make the following illegal:
```haskell
padd :: Term s (PInteger :--> PInteger :--> PInteger)
padd = phoistAcyclic $ plam $ \_ -> perror
```

It would be illegal because then the following would change semantics:
```haskell
padd' = plam \x y -> padd # x # y
```

The reason being that partially applying the function will now cause
an error.

Do note that the following is already "illegal":
```haskell
padd :: Term s (PInteger :--> PInteger :--> PInteger)
padd = phoistAcyclic $ perror
```

In this case, however, it's much less confusing since the script will *always*
err before doing anything, no matter what arguments you pass it.

In addition, the following would also complicate it:
```haskell
pid :: Term s (a :--> a)
pid = phoistAcyclic $ plam $ \x -> x

f :: Term s ((a :--> b :--> c) :--> a :--> b :--> c)
f = plam $ \x y z -> pid # x # y # z

g :: Term s (PInteger :--> PInteger :--> PInteger)
g = plam $ \_ -> perror

-- This will err if `f` is eta-reduced.
h = f # g # 1
```

The above is true even if we have the restriction on hoisted terms.

In the end I think the best possible solution is to do some analysis after the fact,
where we keep track of arities of functions and built-ins.

`plamReduce` is the other possible approach, and is a stop-gap, as the above issue
would be avoided unless users explicitly opt into it, but it's very hard to know
when it's safe to use on a function `f`, as you need to know how `f` is implemented.
In addition I'm not sure how much easier it would be to implement.

Reply:
For the time being we probably want to recommend Plutarch-level functions over Haskell-level functions in general.

Reply:
Plutarch-level functions can also have inefficiencies, but they can be fixed by inlining. E.g. if we have a function that does `\x -> x + x`, then applying it to a variable would do `(\x -> x + x) y`, but if `y` is already small enough then we should just in-line it. We only get this opportunity though if the function *isn't* hoisted.

Reply:
Made the `PIntegral` functions Plutarch level. Also, you should probably put your comments here onto an issue or similar where its more accessible to the general public.